src/components/Admin/Trainer/ControllerDataset.js

```diff
@@ -118,65 +126,111 @@ export class ControllerDataset {
     return idList;
   }
 
-  async getBatch(batchSize, randomness, since) {
-    return new Promise(async resolve => {
-      const batch = {};
-
-      while (Object.keys(batch).length < batchSize) {
-        await this.db
-          .collection('training_data')
-          .orderBy('random')
-          .orderBy('timestamp')
-          .startAt(Math.random(), since)
-          .limit(batchSize * randomness)
-          .get()
-          .then(snapshot => {
-            snapshot.forEach(doc => {
-              batch[doc.id] = doc.data();
-            });
-          });
-      }
-
-      resolve(Object.values(batch).splice(0, batchSize));
+  async fetchData(count, type, onlyLabel) {
+    let ref = this.db
+      .collection('training_data')
+      .where('trusted', '==', true)
+      .orderBy('random');
+    if (onlyLabel) ref = ref.where('label', '==', onlyLabel);
+
+    let random = Math.random();
+    random *=
+      type === 'validation' ? VALIDATION_PERCENTAGE : 1 - VALIDATION_PERCENTAGE;
+    random += type === 'validation' ? 0 : VALIDATION_PERCENTAGE;
+
+    const data = {};
+    const snapshot = await ref
+      .startAt(random)
+      .limit(count)
+      .get();
+    snapshot.forEach(doc => {
+      const {activation, label} = doc.data();
+      data[doc.id] = {activation, label};
     });
+
+    const dataCount = Object.keys(data).length;
+    if (dataCount >= count) {
+      return data;
+    } else {
+      return {
+        ...data,
+        ...(await this.fetchData(count - dataCount, type, onlyLabel))
+      };
+    }
   }
 
-  async getTensors(setSize = 200, randomness = 0.1, since = 0) {
-    const batch = await this.getBatch(setSize, randomness, since);
+  async getTensors(
+    setSize = 200,
+    randomness = 1,
+    dataType = 'training',
+    completion
+  ) {
+    let batch = {};
     const classes = await this.getClasses();
+
+    if (dataType === 'validation') {
+      for (const c in classes) {
+        batch = {
+          ...batch,
+          ...(await this.fetchData(VALIDATION_COUNT, dataType, classes[c]))
+        };
+        completion((c + 1) / classes.length);
+      }
+    } else {
+      while (Object.keys(batch).length < setSize) {
```

Comment:
could get the length once per iteration instead of 2x:
```
 let batchSize = Object.keys(batch).length;
      while (batchSize < setSize) {
        batch = {
          ...batch,
          ...(await this.fetchData(
            Math.min(Math.floor(setSize / randomness), setSize - batchSize),
            dataType
          ))
        };
        batchSize = Object.keys(batch).length;
        completion(batchSize / setSize);
      }
```